File: csharp-workshop/Part4.Behavior.Tests/Service/DiscountCodeServiceTest.cs
using Part4.Behavior.Domain;
using Part4.Behavior.Service;
using Part4.Behavior.Tests.Helper;
using Xunit;

namespace Part4.Behavior.Tests.Service;

public class DiscountCodeServiceTest
{
    private readonly InMemoryDiscountCodeRepository _discountCodeRepository;
    private readonly DiscountCodeService _discountCodeService;

    public DiscountCodeServiceTest()
    {
        _discountCodeRepository = new InMemoryDiscountCodeRepository();
        _discountCodeService = new DiscountCodeService(_discountCodeRepository);
    }

    [Fact]
    public void P4_CheckDiscountCode_ReturnsTrue_WhenCodeHasAlreadyBeenUsed()
    {
        _discountCodeRepository.WithUsedCode("CUST-001", DiscountCode.SUMMER20);

        Assert.True(_discountCodeService.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20));
    }

    [Fact]
    public void P4_CheckDiscountCode_ReturnsFalse_WhenCodeHasNotBeenUsed()
    {
        Assert.False(_discountCodeService.CheckDiscountCode("CUST-001", DiscountCode.SUMMER20));
    }

    [Fact]
    public void P4_MarkAsUsed_MakesCodeCheckedAsUsed()
    {
        _discountCodeService.MarkAsUsed("CUST-001", DiscountCode.SUMMER20);

        Assert.True(_discountCodeRepository.IsMarkedAsUsed("CUST-001", DiscountCode.SUMMER20));
    }
}

File: typescript-workshop/part-4-behavior/tests/service/DiscountCodeServiceTest.ts
import { describe, it, expect } from 'vitest';
import { DiscountCode } from '../../src/domain/DiscountCode';
import { DiscountCodeService } from '../../src/service/DiscountCodeService';
import { InMemoryDiscountCodeRepository } from '../helper/InMemoryDiscountCodeRepository';

describe('DiscountCodeService', () => {
  it('p4_checkDiscountCode_returnsTrue_whenCodeHasAlreadyBeenUsed', () => {
    const repository = new InMemoryDiscountCodeRepository()
      .withUsedCode('CUST-001', DiscountCode.SUMMER20);
    const discountCodeService = new DiscountCodeService(repository);

    expect(discountCodeService.checkDiscountCode('CUST-001', DiscountCode.SUMMER20)).toBe(true);
  });

  it('p4_checkDiscountCode_returnsFalse_whenCodeHasNotBeenUsed', () => {
    const repository = new InMemoryDiscountCodeRepository();
    const discountCodeService = new DiscountCodeService(repository);

    expect(discountCodeService.checkDiscountCode('CUST-001', DiscountCode.SUMMER20)).toBe(false);
  });

  it('p4_markAsUsed_storesCodeInRepository', () => {
    const repository = new InMemoryDiscountCodeRepository();
    const discountCodeService = new DiscountCodeService(repository);

    discountCodeService.markAsUsed('CUST-001', DiscountCode.SUMMER20);

    expect(repository.isMarkedAsUsed('CUST-001', DiscountCode.SUMMER20)).toBe(true);
  });
});
